[SNAP-3110] Changes for SNAP-3110 with fix for backward compatibility

vatsalmevada opened the pull request:
## Changes proposed in this pull request

Re-merging changes for SNAP-3110 with the fix for backward compatibility.

Earlier we introduced a new parameter for the procedure
`CREATE_OR_DROP_RESERVOIR_REGION`. This can break the upgrade from older
versions while DDL replay. Instead of exposing extra parameter for removal of sampler,
we are not removing it always along with the reservoir region.

Fixing failing test `AQPCatalogConsistencyDUnitTest`

## Patch testing
- `AQPCatalogConsistencyDUnitTest` passing
- will trigger `precheckin -Pstore`

## Is precheckin with -Pstore clean?
yet to be run

## Other PRs 

https://github.com/SnappyDataInc/snappydata/pull/1423
https://github.com/SnappyDataInc/snappy-aqp/pull/194

Commit:
Removing reservoir region before attempting to
drop sample table while fixing inconsistent
catalog.

Commit:
- Removing cached sampler for the sample table while fixing catalog
inconsistency.
- Adding DUnit tests.

Commit:
Merge branch 'snappy/master' into SNAP-3110

Commit:
- added a dunit test for lead stopped scenario
- incorporating review comments

Commit:
Merge branch 'snappy/master' into SNAP-3110

Commit:
- To maintain backward compatibility exposing new procedure
 CREATE_OR_DROP_RESERVOIR_REGION_V2 instead of adding additional
  parameter in existing procedure CREATE_OR_DROP_RESERVOIR_REGION.
- Cleaning cached sampler entries even from the where reservoir region
doesn't exist.

vatsalmevada commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -1584,11 +1593,61 @@ public static void CREATE_OR_DROP_RESERVOIR_REGION(String reservoirRegionName,
+   *                      along with the region. This flag will be a ignored if the `isDrop`
+   *                      flag is false.
+   */
+  public static void CREATE_OR_DROP_RESERVOIR_REGION_V2(String reservoirRegionName,
```
Instead of `CREATE_OR_DROP_RESERVOIR_REGION_V2`, we considered keeping the stored procedure name `DROP_RESERVOIR_REGION_AND_REMOVE_SAMPLER` and keep the behavior of this new procedure accordingly. We dropped this approach as by that approach conflation may not work.

Note that the conflation won't work with this approach also for the regions created on older versions of the product.

@sumwale, @kneeraj, @ahshahid please share your thoughts on this.

sumwale replied:
Right, that is where one of the other two approaches can be used. However, I think we can live with lack of conflation for this case which should not be anything more than some wasted effort.

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/db/FabricDatabase.java:
```diff
@@ -857,8 +858,18 @@ private static final void dropTables(EmbedConnection embedConn,
+          Region<Object, Object> reservoirRegion = Misc.getRegionForTable(reservoirRegionName,
+              false);
+          boolean tableContainsData = (columnBuffer.size() != 0) || (rowBuffer != null && rowBuffer.size() != 0)
+              || (reservoirRegion != null && reservoirRegion.size() != 0);
```
I think size check should be removed. The reservoir could have been completely rolled into column store and may not have anything but we still want to remove the region. You can check for !isDestroyed()

@dshirish Not sure why size checks are there for columnBuffer/rowBuffer regions. Either could be zero sized but they need to be cleaned from DataDictionary regardless. Could test for !isDestroyed()

dshirish replied:
@sumwale This was just to be on safe side to avoid dropping tables that contain data etc. The repair_catalog() proc by default does not drop a table if columnBuffer / rowBuffer contains data.  The procedure accepts a boolean parameter,  `removeTablesWithData` that allows removing the tables if they contain data. If this is false, just logs a warning.  Could be removed if this is not of much use.

sumwale replied:
@dshirish This method will only be called if "removeTablesWithData" is true, right? In that case its quite possible that rowBuffer table size is 0 because all data has folded into column store, so then this check will skip deleting this region. Won't that cause problems when user tries to recreate the table failing with region already exists?

dshirish replied:
@sumwale The method (FabricDatabase#dropTables) is always called even if "removeTablesWithData" is false.  The "tableContainsData" flag is set after checking sizes of both columnBuffer and rowBuffer and will skip dropping it if either of the size is non zero and will proceed to drop the table only if user has specified so using "removeTablesWithData". Hope I am following the question.

dshirish replied:
@sumwale If column store is skipped "columnBatchTableExists" flag is set to true to skip dropping rowBuffer as well.

sumwale replied:
ok

sumwale commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -1584,11 +1593,61 @@ public static void CREATE_OR_DROP_RESERVOIR_REGION(String reservoirRegionName,
+      throw TransactionResourceImpl.wrapInSQLException(t);
+    }
+  }
+
```
Why code duplication? CREATE_OR_DROP_RESERVOIR_REGION should invoke CREATE_OR_DROP_RESERVOIR_REGION_V2 with last argument as true. In fact don't quite understand when would that be passed as false? Looks like that argument should always be true in which case it is not required at all, so the original procedure signature can remain as is without backward compatibility concerns.

vatsalmevada replied:
@sumwale, the `removeSampler` argument is passed false from here: `org.apache.spark.sql.sampling.ColumnFormatSamplingRelation#destroy`. This method does clean up of cached sampler separately and we wanted to keep this code untouched.
We can refactor this code and keep using `CREATE_OR_DROP_RESERVOIR_REGION` with remove sampler code executing always if this refactoring looks safe. OR we can keep this code as it is, in that case, the finally block of `destroy` method will be redundant.
Let me know whether this refactoring seems safe. I will rerun precheckin on the changes.

The code duplication was done to use different messages for both the stored procedures as I thought that would be needed for backward compatibility. Maybe my understanding was incorrect.

sumwale replied:
> This method does clean up of cached sampler separately and we wanted to keep this code untouched.

Why?? Isn't it redundant and simply passing the "removeSampler" flag as true should be all that is required. Besides it will also test that new code a bit more.

> OR we can keep this code as it is, in that case, the finally block of destroy method will be redundant.

Yes, simply make this flag as always true and remove the finally block there. In other words remove the CREATE_OR_DROP_RESERVOIR_REGION_V2 and just have the original three argument CREATE_OR_DROP_RESERVOIR_REGION which internally will always do removeSampler. This way nothing related to backward compatibility is required and the code changes will be much simplified.

vatsalmevada replied:
@sumwale I have committed the suggested changes and will trigger precheckin soon.

While doing changes I noticed that by removing the finally block of `destroy` method, the remove sampler flow will be called from inside the procedure which will be called only when `isReservoirAsRegion`(which is set by `spark.sql.aqp.reservoirAsRegion`) flag is true. Earlier in the finally block the remove sampler flow was executed irrespective of this property. Does this seem appropriate?

sumwale replied:
Ok, got it. The sampler removal block is required for non-region case too, so the finally block should be retained under if (!isReservoirAsRegion) condition. Afaik the API based sampling call on a DataFrame does not using a region rather just the in-memory route where isReservoirRegion is false.

sumwale reviewed (changes requested):
See review comments. Specifically I don't understand when would "removeSampler" argument be false for the procedure. As far as I can tell it should always be true in which case you don't need it or the V2 procedure at all.

Commit:
Incorporating review comments-
Updating the behavior of `CREATE_OR_DROP_RESERVOIR_REGION` to always
clean cached sampler entry eliminating the need to new procedure
`CREATE_OR_DROP_RESERVOIR_REGION_V2` altogether. Refactored
`ColumnFormatSamplingRelation#destroy` accordingly.

Commit:
Merge branch 'snappy/master' into SNAP-3110

sumwale reviewed (approved):
LGTM

Commit:
minor code formatting
